src/org/daisy/dotify/formatter/impl/obfl/OBFLExpressionBase.java

```diff
@@ -4,118 +4,161 @@
         this.ef = ef;
         this.exp = exp;
-        this.pageNumberVariable = DEFAULT_PAGE_NUMBER_VARIABLE_NAME;
-        this.volumeNumberVariable = DEFAULT_VOLUME_NUMBER_VARIABLE_NAME;
-        this.volumeCountVariable = DEFAULT_VOLUME_COUNT_VARIABLE_NAME;
-        this.sheetCountVariable = DEFAULT_SHEET_COUNT_VARIABLE_NAME;
-        this.volumeSheetCountVariable = DEFAULT_VOLUME_SHEET_COUNT_VARIABLE_NAME;
-        if (extended) {
-            this.metaVolumeNumberVariable = DEFAULT_EVENT_VOLUME_NUMBER;
-            this.metaPageNumberVariable = DEFAULT_EVENT_PAGE_NUMBER;
-        } else {
-            this.metaVolumeNumberVariable = null;
-            this.metaPageNumberVariable = null;
-        }
-    }
-
-    public void setPageNumberVariable(String pageNumberVariable) {
-        if (pageNumberVariable == null) {
-            this.pageNumberVariable = DEFAULT_PAGE_NUMBER_VARIABLE_NAME;
-        } else {
-            this.pageNumberVariable = pageNumberVariable;
-        }
-    }
-
-    public void setVolumeNumberVariable(String volumeNumberVariable) {
-        if (volumeNumberVariable == null) {
-            this.volumeNumberVariable = DEFAULT_VOLUME_NUMBER_VARIABLE_NAME;
-        } else {
-            this.volumeNumberVariable = volumeNumberVariable;
-        }
-    }
-
-    public void setVolumeCountVariable(String volumeCountVariable) {
-        if (volumeCountVariable == null) {
-            this.volumeCountVariable = DEFAULT_VOLUME_COUNT_VARIABLE_NAME;
-        } else {
-            this.volumeCountVariable = volumeCountVariable;
+        this.variables = new HashSet<>();
+        for (OBFLVariable v : variables) {
+            switch (v) {
+            case PAGE_NUMBER:
+            case VOLUME_NUMBER:
+            case VOLUME_COUNT:
+            case STARTED_VOLUME_NUMBER:
+                break;
+            case STARTED_PAGE_NUMBER:
+                if (this.variables.contains(OBFLVariable.STARTED_VOLUME_FIRST_CONTENT_PAGE_NUMBER)) {
+                    throw new IllegalArgumentException(
+                        "STARTED_PAGE_NUMBER and STARTED_VOLUME_FIRST_CONTENT_PAGE_NUMBER " +
+                        "may not both be used in the same expression.");
+                }
+                break;
+            case STARTED_VOLUME_FIRST_CONTENT_PAGE_NUMBER:
+                if (this.variables.contains(OBFLVariable.STARTED_PAGE_NUMBER)) {
+                    throw new IllegalArgumentException(
+                        "STARTED_PAGE_NUMBER and STARTED_VOLUME_FIRST_CONTENT_PAGE_NUMBER " +
+                        "may not both be used in the same expression.");
+                }
+                break;
+            case SHEET_COUNT:
+            case VOLUME_SHEET_COUNT:
+                break;
+            default:
+                throw new IllegalArgumentException(); // coding error
+            }
+            this.variables.add(v);
+            setVariableName(v, null);
```

Comment:
@bertfrees my IDE (NetBeans) warns me against an "Overridable method call in constructor".
I will fix this.

For more details, please see: https://stackoverflow.com/questions/3404301/whats-wrong-with-overridable-method-calls-in-constructors

Reply:
RIght good point.